File: 1073-number-of-enclaves/number-of-enclaves.cpp
class Solution {
public:
    int rows;
    int cols;

    bool dfs(int row, int col, vector<vector<int>>& grid, int& cells) {
        if(row < 0 || row >= rows || col < 0 || col >= cols) {
            return false;
        }
        if(grid[row][col] == 0) {
            return true;
        }
        cells++;
        grid[row][col] = 0;

        bool down = dfs(row + 1, col, grid, cells);
        bool up = dfs(row - 1, col, grid, cells);
        bool right = dfs(row, col + 1, grid, cells);
        bool left = dfs(row, col - 1, grid, cells);

        return down && up && right && left;
    }

    int numEnclaves(vector<vector<int>>& grid) {
        rows = grid.size();
        cols = grid[0].size();

        int totalCells = 0;
        for(int i = 0; i < rows; i++) {
            for(int j = 0; j < cols; j++) {
                if(grid[i][j] == 1) {
                    int cells = 0;
                    if(dfs(i, j, grid, cells)) 
                        totalCells += cells;
                }
            }
        }
        return totalCells;
    }
};

File: 1073-number-of-enclaves/number-of-enclaves.java
class Solution {
    public int rows, cols;

    public void DFS(int i, int j, int[][] grid) {
        if(i < 0 || i >= rows || j < 0 || j >= cols || grid[i][j] != 1)
            return;
        
        grid[i][j] = 0;
        DFS(i + 1, j, grid);
        DFS(i - 1, j, grid);
        DFS(i, j + 1, grid);
        DFS(i, j - 1, grid);
    }

    public int numEnclaves(int[][] grid) {
        this.rows = grid.length;
        this.cols = grid[0].length;

        for(int i = 0; i < cols; i++) {
            if(grid[0][i] == 1)
                DFS(0, i, grid);
            if(grid[rows - 1][i] == 1) 
                DFS(rows - 1, i, grid);
        }

        for(int i = 0; i < rows; i++) {
            if(grid[i][0] == 1) 
                DFS(i, 0, grid);
            if(grid[i][cols - 1] == 1)
                DFS(i, cols - 1, grid);
        }

        int landCells = 0;
        for(int i = 0; i < rows; i++) {
            for(int j = 0; j < cols; j++) {
                if(grid[i][j] == 1) 
                    landCells++;
            }
        }
        return landCells;
    }
}